lagom/integrations/django.py

```diff
@@ -0,0 +1,145 @@
+"""
+Django integration.
+
+Usage:
+
+```
+# urls.py
+from django.urls import path
+
+from . import views
+from .dependency_config import container
+
+
+urlpatterns = [
+    path('', container.view(views.index), name='index'),
+    path('extra', container.view(views.CBVexample), name='extra')
+]
+```
+
+```
+#views.py
+
+def index(request, dep: SomeDep, questions: DjangoModel[Question]):
+    new_question = questions.new(question_text="What's next?", pub_date=timezone.now())
+    new_question.save()
+    return HttpResponse(f"plain old function: {dep.message} with {questions.objects.all().count()} questions")
+
+
+class CBVexample(View):
+    def get(self, request, dep: SomeDep):
+        return HttpResponse(f"Class based: {dep.message}")
+
+```
+
+```
+# dependency_config.py
+# a per app dep injection container
+
+# Lists all models that should be available
+container = DjangoContainer(models=[Question], request_singletons= [SomeCache])
+container[SomeService] = SomeService("connection details etc")
+```
+
+"""
+import types
+from typing import TypeVar, Generic, List, Type, Optional
+
+from django.db.models import Manager, Model
+from django.views import View
+
+from .. import Container
+
+M = TypeVar("M", bound=Model)
+
+
+class DjangoModel(Generic[M]):
+    """Wrapper around a django model for injection hinting
+    Usage:
+        container = DjangoContainer(models=[Question])
+
+        @bind_to_container(container)
+        def load_first_question(questions: DjangoModel[Question]):
+            return questions.objects.first()
+
+    """
+
+    model: Type[M]
+
+    def __init__(self, model: Type[M]):
+        """
+
+        :param model: The django model class
+        """
+        self.model = model
+
+    @property
+    def objects(self) -> Manager:
+        """ Equivalent to MyModel.objects
+
+        :return:
+        """
+        return self.model.objects  # type: ignore
```

Comment:
This could possibly become a hassle with multiple Manager references. That said, I haven't found manager methods named other than "objects" and/or having multiple, in themselves very useful. To solve that I guess you would have to introspect for Manager inheritance on all attributes.

Reply:
This is a really good point. I'd not considered multiple managers. I guess this covers 90% (number entirely made up) of usage.

🤔 introspection might be a nice fit. Or alternatively something "trust based" so that `model.managers.anything` works.